### PR TITLE
DateRange mereology/lattice logic

### PR DESCRIPTION
**Pull request by wroberts**
Hi @shx2,

First of all, thank you for this sweet library!  Really nicely done.

I'm sending a pull request which may or may not interest you, adding a bunch of methods to the DateRange type.

I've been looking for a library to perform "arithmetic" on date ranges (figuring out which date ranges overlap with each other, computing the set unions, intersections and differences of their constituent days, etc.) and so far I haven't found anything.  On the other hand, I did come across your `yyyymmdd` module, which is pretty neat.  So I just decided to add the date range operations into `yyyymmdd`, as it solves my particular problem for now.  I'm opening this PR here in case you'd like to merge these changes back.

The basic idea is that date range objects form a kind of [mereology](https://en.wikipedia.org/wiki/Mereology) or lattice structure: date ranges can be parts of other date ranges, or they can overlap, or be "adjacent" to each other. For instance, Monday to Friday (inclusive) is adjacent to the weekend (inclusive), but doesn't overlap it.  The "underlap" relation means either overlapping or adjacency.  In cases where two date ranges overlap, it's meaningful to define the intersection (which you already have), as well as the difference.  In cases where two date ranges underlap, it's also meaningful to define the union.

```
# jan 3 to jan 6 is not a part of jan 1 to jan 5
>>> DateRange(20000101, 20000105) >= DateRange(20000103, 20000106)
False
# jan 3 to jan 5 IS a part of jan 1 to jan 5
>>> DateRange(20000101, 20000105) >= DateRange(20000103, 20000105)  
True
# jan 1 to jan 5  overlaps with jan 3 to jan 6
>>> DateRange(20000101, 20000105).overlaps(DateRange(20000103, 20000106))
True
# jan 1 to jan 5  "minus" jan 3 to jan 6 is jan 1 to jan 3
>>> DateRange(20000101, 20000105) - DateRange(20000103, 20000106)  
[DateRange(20000101, 20000103)]
# jan 1 to jan 5  "union" jan 3 to jan 6 is jan 1 to jan 6
>>> DateRange(20000101, 20000105) | DateRange(20000103, 20000106)  
DateRange(20000101, 20000106)
```

I've added test cases to cover all the new code.

There are some quirks to this stuff.  The new methods don't work when `step` is not 1; I think it makes sense to talk about the intersection of two date ranges with non-unit stepsize, as you've done, but not to talk about the union or difference.  Another weirdness is that the difference operator returns a list, because the difference can be either empty, non-empty, or two non-empty date ranges:

```
# jan 1 to jan 10 "minus" jan 3 to jan 6 is jan 1 to jan 3 AND jan 6 to jan 10
>>> DateRange(20000101, 20000110) - DateRange(20000103, 20000106)  
[DateRange(20000101, 20000103), DateRange(20000106, 20000110)]
```

OK, that's about it.  Let me know if you're interested.  I'd be happy to copy-paste some of this stuff into the README if you'd like.

Best, Will

### CODE REVIEW

**Comment by shx2** [comment]
@wroberts Thank you for contributing. Much appreciated!
I will take a look at it soon.

File: yyyymmdd/range.py [hunk 1]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+            )
+        if not bool(self._ordrange):
+            return True
+        return (
+            (other._ordrange.start <= self._ordrange.start)
+            and (self._ordrange.stop <= other._ordrange.stop)
+            and (self._ordrange != other._ordrange)
+        )
+
+    def overlaps(self, other):
```
Comment (shx2) [inline yyyymmdd/range.py:415]: Wouldn't it make sense to implement this method as `bool(self & other )`?
This way, it also supports non-unit step.

File: yyyymmdd/range.py [hunk 2]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+        return self.is_adjacent(other) or self.overlaps(other)
+
+    def __or__(self, other):
+        return self.union(other)
+
+    def union(self, other):
+        """
+        Computes the union of this DateRange and the given DateRange other.
+
+        The union of X and Y is only defined if X and Y underlap.
```
Comment (shx2) [inline yyyymmdd/range.py:440]: Add to docstring the fact only step=1 is supported

File: yyyymmdd/range.py [hunk 3]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+        return (bool(self._ordrange) and bool(other._ordrange)) and (
+            (other._ordrange.start <= self._ordrange.start < other._ordrange.stop)
+            or (self._ordrange.start <= other._ordrange.start < self._ordrange.stop)
+        )
+
+    def underlaps(self, other):
+        """
+        Finds the underlap of this DateRange and the given DateRange other.
+
+        The X and Y underlap if X and Y overlap or are adjacent.
```
Comment (shx2) [inline yyyymmdd/range.py:427]: Add to docstring the fact only step=1 is supported

File: yyyymmdd/range.py [hunk 4]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+            (other._ordrange.start <= self._ordrange.start)
+            and (self._ordrange.stop <= other._ordrange.stop)
+            and (self._ordrange != other._ordrange)
+        )
+
+    def overlaps(self, other):
+        """
+        Tests whether this DateRange overlaps with the given DateRange other.
+
+        X and Y overlap if there is a part of X that is also a part of Y.
```
Comment (shx2) [inline yyyymmdd/range.py:419]: Add to docstring the fact only step=1 is supported

File: yyyymmdd/range.py [hunk 5]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+        return (self._ordrange.stop == other._ordrange.start) or (
+            other._ordrange.stop == self._ordrange.start
+        )
+
+    def is_proper_part(self, other):
+        """
+        Tests whether this DateRange is a proper part of (is fully
+        contained in) the given DateRange other.
+
+        X is a proper part of Y if X is a part of Y and Y is not a part of X.
```
Comment (shx2) [inline yyyymmdd/range.py:403]: Add to docstring the fact only step=1 is supported

File: yyyymmdd/range.py [hunk 6]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+        )
+
+    def __ge__(self, other):
+        return other <= self
+
+    def is_adjacent(self, other):
+        """
+        Tests whether this DateRange is adjacent to (externally connected)
+        to the given DateRange other.
+        """
```
Comment (shx2) [inline yyyymmdd/range.py:392]: Add to docstring the fact only step=1 is supported

File: yyyymmdd/range.py [hunk 7]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+            if sign > 0
+            else stop_max_inv(self.stop, other.stop)
+        )
         return type(self)(start, stop, step)
 
+    # ===========================================================================
+    # mereological part structure
+    # ===========================================================================
+
+    def __lt__(self, other):
```
Comment (shx2) [inline yyyymmdd/range.py]: I strongly prefer overriding the order-operators only when an order relation is well-defined and intuitive.
Otherwise, it leads to "surprises", e.g. `a<=b and b<=a` does not imply `a==b`, or `a<b or a>b or a==b is False`.
Please delete the operators.  The `is_proper_part` method suffices.

File: yyyymmdd/range.py [hunk 8]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+            part2 = DateRange(
+                self._o2d(other._ordrange.stop), self._o2d(self._ordrange.stop)
+            )
+            if bool(part1):
+                if bool(part2):
+                    return [part1, part2]
+                else:
+                    return [part1]
+            else:
+                if bool(part2):
```
Comment (shx2) [inline yyyymmdd/range.py]: Should avoid `if bool(...): ...` (in other places too).

```suggestion
            if part1:
                if part2:
                    return [part1, part2]
                else:
                    return [part1]
            else:
                if part2:
```

File: yyyymmdd/range.py [hunk 9]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+                    return [part2]
+                return [DateRange.empty()]
```
Comment (shx2) [inline yyyymmdd/range.py]: Why return a list with an empty range? An empty list is just as good.
```suggestion
                return []
```

File: tests/test_range.py [hunk 10]
```diff
@@ -2,10 +2,13 @@
 Unit-tests for DateRange.
 """
 
-import unittest
```
Comment (shx2) [inline tests/test_range.py:5]: If you wish to make styling changes, that's fine, but please submit that as a separate PR.

File: yyyymmdd/range.py [hunk 11]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+
+    def __gt__(self, other):
+        return other.is_proper_part(self)
+
+    def __le__(self, other):
+        if self.step != 1 or other.step != 1:
+            raise ValueError(
+                "Mereological methods are undefined for ranges with steps != 1: %r & %r"
+                % (self, other)
+            )
```
Comment (shx2) [inline yyyymmdd/range.py]: code duplication. please move to a method.

File: yyyymmdd/range.py [hunk 12]
```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
...
+
+    def __lt__(self, other):
+        return self.is_proper_part(other)
+
+    def __gt__(self, other):
+        return other.is_proper_part(self)
+
+    def __le__(self, other):
+        if self.step != 1 or other.step != 1:
+            raise ValueError(
```
Comment (shx2) [inline yyyymmdd/range.py]: I wonder if `NotImplementedError` is more suitable here, because for (at least) some of the mereology operations, we _could_ implement them.
I'm fine with leaving it unimplemented and raising `NotImplementedError`, until someone decides it is worth the effort to do the extra work.

**Review by shx2 — CHANGES_REQUESTED** [review]
Good work! This addition adds real value to this project.

Please don't be disheartened by the many comments... Only a few are actually substantial. The rest are just me being pedantic.

Also, I apologize for taking so long to review this.

**Comment by wroberts** [comment]
Life has got in the way of progressing on this PR, but I found some time today. Most change requests have been adressed.

**Comment by wroberts** [comment]
ValueError changed out for NotImplementedError.

File: tests/test_range.py [hunk 13]
```diff
@@ -149,5 +150,157 @@ def test_pickle(self):
             self.assertEqual(r1, r2)
             self.assertEqual(type(r1), type(r2))
 
+    def test_mereology(self):
+        for _ in range(1000):
+            d1 = DateRange.DATE_TYPE.fromordinal(random.randint(0, 1000000))
```
Comment (shx2) [inline tests/test_range.py:156]: unittests should be deterministic.
However, you may use `random`, as long as you initialize it with a hardcoded seed (to make it deterministic).

**Review by shx2 — CHANGES_REQUESTED** [review]
Awesome.
Just one last minor comment.

File: tests/test_range.py (hunk 13 above)
Comment (wroberts) [reply tests/test_range.py:156]: excellent idea. done.

**Comment by shx2** [comment]
Thank you for contributing!
